experiments/bert/bert.py

```diff
@@ -0,0 +1,231 @@
+import math
+
+import numpy as np
+import pandas as pd
+import torch
+from pytorch_pretrained_bert import BertTokenizer, BertForSequenceClassification
+from pytorch_pretrained_bert.optimization import BertAdam
+from torch.nn.utils import clip_grad_norm_
+from torch.optim import Optimizer
+from torch.optim.optimizer import required
+from tqdm import tqdm
+
+from transfer_nlp.loaders.loaders import DatasetSplits, DatasetHyperParams, DataFrameDataset
+from transfer_nlp.loaders.vectorizers import Vectorizer
+from transfer_nlp.loaders.vocabulary import Vocabulary
+from transfer_nlp.plugins.config import register_plugin
+
+tqdm.pandas()
+
+
+@register_plugin
+class BertVectorizer(Vectorizer):
+    def __init__(self, data_file: str):
+        super().__init__(data_file=data_file)
+        self.tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')
+
+        df = pd.read_csv(data_file)
+        target_vocab = Vocabulary(add_unk=False)
+        for category in sorted(set(df.category)):
+            target_vocab.add_token(category)
+        self.target_vocab = target_vocab
+
+    def vectorize(self, title: str, max_seq_length: int) -> np.array:
+        tokens = self.tokenizer.tokenize(title)
+        tokens = ["[CLS]"] + tokens + ["[SEP]"]
+        token_type_ids = [0] * len(tokens)
+        input_ids = self.tokenizer.convert_tokens_to_ids(tokens)
+        attention_mask = [1] * len(input_ids)
+        padding = [0] * (max_seq_length - len(input_ids))
+        input_ids += padding
+        attention_mask += padding
+        token_type_ids += padding
+
+        return np.array(input_ids), np.array(attention_mask), np.array(token_type_ids)
+
+
+@register_plugin
+class BertDataset(DatasetSplits):
+
+    def __init__(self, data_file: str, batch_size: int, dataset_hyper_params: DatasetHyperParams):
+        self.df = pd.read_csv(data_file)
+        # np.random.shuffle(self.df.values)  # Use this code in dev mode
+        # N = 100
+        # self.df = self.df.head(n=N)
+
+        # preprocessing
+        self.vectorizer: Vectorizer = dataset_hyper_params.vectorizer
+
+        self.max_sequence = 0
+        for title in tqdm(self.df.title, desc="Getting max sequence"):
+            tokens = self.vectorizer.tokenizer.tokenize(text=title)
+            self.max_sequence = max(self.max_sequence, len(tokens))
+        self.max_sequence += 2
+
+        self.df['x_in'] = self.df['title'].progress_apply(lambda x: self.vectorizer.vectorize(title=x, max_seq_length=self.max_sequence))
```

Comment:
it might be cleaner not to add `x_in` to the data frame. instead you could just save it as a local variable and then only add the actual parameter columns?

Reply:
right